File: Games/NoughtsAndCrosses/src/main/java/jokatu/game/games/noughtsandcrosses/game/NoughtsAndCrossesGame.java
package jokatu.game.games.noughtsandcrosses.game;

import jokatu.game.Game;
import jokatu.game.GameID;
import jokatu.game.games.noughtsandcrosses.player.NoughtsAndCrossesPlayer;
import jokatu.game.stage.GameOverStage;
import jokatu.game.stage.JoiningStage;
import jokatu.game.stage.machine.SequentialStageMachine;
import jokatu.game.status.StandardTextStatus;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;

public class NoughtsAndCrossesGame extends Game<NoughtsAndCrossesPlayer> {

	public static final String NOUGHTS_AND_CROSSES = "Noughts and crosses";

	private final StandardTextStatus status = new StandardTextStatus();

	NoughtsAndCrossesGame(GameID identifier) {
		super(identifier);

		stageMachine = new SequentialStageMachine(Arrays.asList(
				() -> new JoiningStage<>(NoughtsAndCrossesPlayer.class, players, 2, status),
				() -> new AllegianceStage(players.values(), status),
				() -> new InputStage(players.values(), status),
				() -> new GameOverStage(status)
		));

		status.observe(this::fireEvent);
	}

	@NotNull
	@Override
	public String getGameName() {
		return NOUGHTS_AND_CROSSES;
	}
}

File: Games/NoughtsAndCrosses/src/main/kotlin/jokatu/game/games/noughtsandcrosses/game/NoughtsAndCrossesGame.kt
package jokatu.game.games.noughtsandcrosses.game

import jokatu.game.Game
import jokatu.game.GameID
import jokatu.game.games.noughtsandcrosses.player.NoughtsAndCrossesPlayer
import jokatu.game.stage.GameOverStage
import jokatu.game.stage.JoiningStage
import jokatu.game.stage.machine.SequentialStageMachine
import jokatu.game.status.StandardTextStatus

class NoughtsAndCrossesGame internal constructor(identifier: GameID) : Game<NoughtsAndCrossesPlayer>(identifier) {

	private val status = StandardTextStatus()

	override val gameName: String
		get() = NOUGHTS_AND_CROSSES

	init {
		stageMachine = SequentialStageMachine(
				{ JoiningStage(NoughtsAndCrossesPlayer::class, players, 2, status) },
				{ AllegianceStage(players.values, status) },
				{ InputStage(players.values, status) },
				{ GameOverStage(status) }
		)

		status.observe({ this.fireEvent(it) })
	}

	companion object {
		const val NOUGHTS_AND_CROSSES = "Noughts and crosses"
	}
}
